doc/ip/next_hop_group_hld.md

```diff
@@ -0,0 +1,148 @@
+
+# 1 Introduction
+Currently the route table within APP_DB contains all next hop information for a route embedded in that route's entry in the table.
+
+At high scale (particularly when handling millions of routes all routed over multiple next hops) this is inefficient both in terms of performance and occupancy.  A given route entry needs to be built up including all its next hop information, which gets sent to and stored in APP_DB.  Orchagent then needs to receive, parse and handle all the information in that route entry every time.
+
+A more efficient system would involve managing the next hop groups in use by the route table separately, and simply have the route table specify a reference to which next hop group to use.  Since at scale many routes will use the same next hop groups, this requires much smaller occupancy per route, and so more efficient building, transmission and parsing of per-route information.
+
+# 2 Requirement Overview
+## 2.1 Functional Requirements
+
+- Improve efficiency of storage of routing information in APP_DB.
+- Reduce time to reprogram routes when next hop information changes.
+
+## 2.2 Configuration and Management Requirements
+- The existing 'show ip route' and 'show ipv6 route' commands must be supported when using the new next hop group table.
+
+## 2.3 Scalability Requirements
+- Unchanged in requirements, but will improve scalability in terms of storage and programming time.
+
+## 2.4 Warm Boot Requirements
+- Unchanged - the new next hop group table must be compatible with existing warm boot requirements.
+
+# 3 Design
+## 3.1 Overview
+This design directly changes APP_DB and orchagent, along with minor changes to support the existing CLI when using the new APP_DB table.
+
+## 3.2 DB Changes
+### 3.2.1 APP DB
+Currently the ROUTE_TABLE in APP_DB includes all next hop information for the route, as follows:
+```
+### ROUTE_TABLE
+    ;Stores a list of routes
+    ;Status: Mandatory
+    key           = ROUTE_TABLE:prefix
+    nexthop       = *prefix, ;IP addresses separated “,” (empty indicates no gateway)
+    ifname        = *PORT_TABLE.key,   ; zero or more separated by “,” (zero indicates no interface)
+    blackhole     = BIT ; Set to 1 if this route is a blackhole (or null0)
+ ```
+
+This design adds a new NEXT_HOP_GROUP_TABLE, to store next hop group information to be used by one or more routes.
+```
+### NEXT_HOP_GROUP_TABLE
+    ;Stores a list of groups of one or more next hops
+    ;Status: Mandatory
+    key           = NEXT_HOP_GROUP_TABLE:string ; arbitrary identifier for the next hop group
+    nexthop       = *prefix,           ; IP addresses separated “,” (empty indicates no gateway)
+    ifname        = *PORT_TABLE.key,   ; zero or more separated by “,” (zero indicates no interface)
+```
+The ROUTE_TABLE is then extended to allow a reference to the next hop group to be specified, instead of the current nexthop and intf fields.
+```
+### ROUTE_TABLE
```

Comment:
>ROUTE_TABLE [](start = 4, length = 11)

What if application add a route which refers to a nexthop group that hasn’t be created, how to deal with that case?

Reply:
Let's say use add a static route, the nexthop group is not existing in SAI or AppDB table, and there are ASIC resource available. How to create this nexthop group? Waiting does not solve anything.

Reply:
If an application is using this model, then it's the application's responsibility to program the NEXT_HOP_GROUP_TABLE and ROUTE_TABLE - to program just the ROUTE_TABLE would be a bug. Note that the ROUTE_TABLE can still contain the next hop information, if an application chooses to do it that way, and in that case routeorch would manage the next hop group as is currently the case.

Reply:
Discuss under another assumption (not following above), we are using dynamic routes in most cases, we don't change fpmsyncd, "ROUTE_TABLE can still contain the next hop information", everything is like before? I feel the 2-option design is too complex. Can we just design with one option?

Reply:
In due course we'd like to move to 1 option, but removing the existing mechanism would require both changing fpmsyncd (which would be possible), and ensuring that everyone using the current mechanism but not fpmsyncd is able to change to the new mechanism